File: it.unibo.bls19Local/src/main/java/it/unibo/bls/applLogic/BlsApplicationLogic.java
package it.unibo.bls.applLogic;

import it.unibo.bls.interfaces.IControlLed;
import it.unibo.bls.interfaces.ILed;
import it.unibo.bls.utils.Utils;

public class BlsApplicationLogic implements IControlLed { //Note: it is NOT an observer
	private ILed led;
	private int numOfCalls = 0;
	private boolean doBlink = false;
	private final Object monitor = new Object();

 	public void setControlled( ILed led ){ //no controlled => no activity
		this.led = led;
		//doBlinkTheLed();
		doBlinkTheLedWaiting();		//Avoid polling on doBlink (numOfCalls)
	}
	/*
	NAIVE solution based on polling on numOfCalls
	 */
	private void doBlinkTheLed(){
		new Thread(){
			public void run(){
 				System.out.println("	BlsApplicationLogic | doBlinkTheLed Thread STARTS looping ...");
					while (true) {
						if ( doBlink ) switchTheLed(); //else do nothing
						Utils.delay(250);
					}
			}
		}.start();
	}

	private void switchTheLed(){
		if( led == null ) return;		//defensive programming
		if( led.getState() ) led.turnOff(); else led.turnOn();
	}

	@Override //from IControlLed
	public void execute( String cmd  ){
		numOfCalls++;
		doBlink = numOfCalls % 2 != 0;
		System.out.println("	BlsApplicationLogic | numOfCalls=" + numOfCalls + " doBlink=" + doBlink);
		synchronized( monitor ) {	//see doBlinkTheLedWaiting
 			monitor.notify();
		}
	}

 	//Useful for testing
 	public int getNumOfCalls() {
		return numOfCalls;
	}

	/*
	AVOID POLLING
	 */
	private void doBlinkTheLedWaiting(){
		new Thread(){
			public void run(){
 				System.out.println("	BlsApplicationLogic | doBlinkTheLedWaiting Thread STARTS   ..."  );
				try {
						while (true) {
							synchronized( monitor ) {
								while( ! doBlink ) monitor.wait();
							}
							switchTheLed();
							Utils.delay(250);
						}
  				}catch(Exception e){}
			}
		}.start();
	}
}

File: it.unibo.bls19Local/src/main/kotlin/it/unibo/bls/kotlin/applLogic/BlsApplicationLogic.kt
package it.unibo.bls.kotlin.applLogic

import it.unibo.bls.interfaces.IControlLed
import it.unibo.bls.interfaces.ILed
import it.unibo.bls.utils.Utils
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.launch

open class BlsApplicationLogic : IControlLed {
    protected var led: ILed? = null
    protected var numCalls = 0
    protected var doBlink = false
    protected val channel = Channel<Boolean>()

    override fun setControlled(led: ILed) {
        this.led = led
        GlobalScope.launch{
            println("FOR COROUTINE | numOfThreads=${Thread.activeCount()} currentThread=${Thread.currentThread().name}")
            doBlinkTheLedWaiting()
        }
    }

     private suspend fun doBlinkTheLedWaiting() {
         while (true) {
             doBlink = channel.receive()        //See execute
             println("	COROUTINE | received=$doBlink")
             if (doBlink) doBlinkTheLed()
         }
    }

    private fun doBlinkTheLed() {
        GlobalScope.launch {
            while ( doBlink ) {
                switchTheLed()
                Utils.delay(250)
            }
            println("	BlsApplicationLogicKt coroutine | doBlinkTheLed ENDS ...")
        }
    }

    protected fun switchTheLed() {
        if (led == null) return        //defensive programming
        if ( led!!.getState() ) led!!.turnOff() else led!!.turnOn()
    }

    override//from
    fun execute(cmd: String) {
        println("	BlsApplicationLogicKt | execute cmd=$cmd  ")
        applLogic( )
      }
     open fun applLogic( ){
        numCalls++
        doBlink = numCalls % 2 != 0
        println("	BlsApplicationLogicKt | execute numCalls=$numCalls doBlink=$doBlink")
        GlobalScope.launch {
            channel.send(doBlink)
        }
    }
    //Useful for testing
     override fun getNumOfCalls(): Int {
        return numCalls
    }
}
